src/nupic/algorithms/SDRClassifier.cpp

```diff
@@ -21,489 +21,211 @@
-      }
-      else {
-        actValueVector.push_back( actValue[0] );
+  // Check if this is a new category & resize the weights table to hold it.
+  const auto maxCategoryIdx = *max_element(categoryIdxList.begin(), categoryIdxList.end());
+  if( maxCategoryIdx >= numCategories_ ) {
+    numCategories_ = maxCategoryIdx + 1;
+    for( auto & vec : weights_ ) {
+      while( vec.size() < numCategories_ ) {
+        vec.push_back( 0.0f );
       }
     }
   }
 
-  for( auto nSteps = steps_.begin(); nSteps != steps_.end(); ++nSteps )
-  {
-    vector<Real64> &likelihoods = result[ *nSteps ];
-    likelihoods.assign( maxBucketIdx_ + 1, 0.0f );
-
-    for( const auto& bit : patternNZ ) {
-      const Matrix& w = weightMatrix_.at(*nSteps);
-      for( size_t i = 0; i < likelihoods.size(); i++ ) {
-        likelihoods.at(i) += get_(w, bit, i);
-      }
+  // Compute errors and update weights.
+  const vector<Real> error = calculateError_(categoryIdxList, pattern);
+  for( const auto& bit : pattern.getSparse() ) {
+    for(size_t i = 0; i < numCategories_; i++) {
+      weights_[bit][i] += alpha_ * error[i];
     }
-    softmax_( likelihoods.begin(), likelihoods.end() );
   }
 }
 
-vector<Real64> SDRClassifier::calculateError_(const vector<UInt> &bucketIdxList,
-                                              const vector<UInt> patternNZ,
-                                              UInt step) {
+
+// Helper function to compute the error signal in learning.
+std::vector<Real> Classifier::calculateError_(
+                    const std::vector<UInt> &categoryIdxList, const SDR &pattern)
+{
   // compute predicted likelihoods
-  vector<Real64> likelihoods(maxBucketIdx_ + 1, 0);
+  auto likelihoods = infer(pattern);
 
-  for (const auto& bit : patternNZ) {
-    const Matrix& w = weightMatrix_.at(step); //row
-    for(UInt i=0; i< (UInt)likelihoods.size(); i++) {
-      likelihoods[i] += get_(w, bit, i);
-    }
+  // Compute target likelihoods
+  vector<Real> targetDistribution(numCategories_ + 1, 0.0);
+  for( size_t i = 0; i < categoryIdxList.size(); i++ ) {
+    targetDistribution[categoryIdxList[i]] = 1.0 / categoryIdxList.size();
   }
-  softmax_(likelihoods.begin(), likelihoods.end());
-
-  // compute target likelihoods
-  vector<Real64> targetDistribution(maxBucketIdx_ + 1, 0.0);
-  const Real64 numCategories = (Real64)bucketIdxList.size();
-  for (size_t i = 0; i < bucketIdxList.size(); i++)
-    targetDistribution[bucketIdxList[i]] = 1.0 / numCategories;
 
-  NTA_ASSERT(likelihoods.size() == targetDistribution.size());
-  for(size_t i = 0; i < likelihoods.size(); i++) {
+  for( size_t i = 0; i < likelihoods.size(); i++ ) {
     likelihoods[i] = targetDistribution[i] - likelihoods[i];
   }
   return likelihoods;
 }
 
 
-void SDRClassifier::softmax_(const vector<Real64>::iterator begin,
-                             const vector<Real64>::iterator end) {
+void nupic::algorithms::sdr_classifier::softmax(PDF::iterator begin, PDF::iterator end) {
```

Comment:
nit, why not `using namespace nupic::algorithms::sdr_classifier` ?

Reply:
I do have the line `using namespace nupic::algorithms::sdr_classifier` at the top of the file, but for some unknown reason I needed to make the namespace explicit for the compiler to find it.